### PR TITLE
const fn integer operations

### LINKED ISSUE #51267: [tracking issue] const fn: count_zeros, count_ones, leading_zeros, trailing_zeros
* [x] implemented in https://github.com/rust-lang/rust/pull/51299
* [ ] documentation
* [x] stabilization PR in https://github.com/rust-lang/rust/pull/51364 

One thing that I would like to do is to compile-time assert that the log2 of an alignment is less a given size. A simple way to do log2 of a power of 2 is to use `trailing_zeros`. But `trailing_zeros` is not const, so it's not possible to do a compile-time check.

The equivalent `__builtin`s in clang are const (example: https://godbolt.org/g/RxWczn) so llvm should be able to handle that.

However, adding `const` to the functions in `core::num` requires adding `const` to the corresponding intrinsics, and that is not supported (`extern items cannot be 'const'`)

### PR DESCRIPTION
**Pull request by faern**
A follow up to #51171 
Fixes #51267

Makes a lot of the integer methods (`swap_bytes`, `count_ones` etc) `const fn`s. See #51267 for a discussion about why this is wanted and the solution used.

### CODE REVIEW

File: src/test/run-pass/const-endianess.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+#![feature(const_int_ops)]
+
+const BE_U32: u32 = 55u32.to_be();
+const LE_U32: u32 = 55u32.to_le();
+const BE_U128: u128 = 999999u128.to_be();
+const LE_I128: i128 = -999999i128.to_le();
+
+fn main() {
+    assert_eq!(BE_U32, 55u32.to_be());
```
Comment (oli-obk) [inline src/test/run-pass/const-endianess.rs]: To ensure that the right hand side isn't const evaluated you need to pass the literal through a dummy function that isn't const fn and is also `#[inline(never)]`

Comment (joshtriplett) [reply src/test/run-pass/const-endianess.rs]: @oli-obk Doesn't the codebase have a dedicated black-boxing function for that purpose?

Comment (oli-obk) [reply src/test/run-pass/const-endianess.rs]: It does, I think it's `test::black_box`, not sure if that's available in run-pass tests.

Comment (faern) [reply src/test/run-pass/const-endianess.rs]: Fixed. I was able to use `test::black_box`.

Btw. Do you think this test is correctly placed? I first put it under ui to just make sure it compiled, never ran any assertions. Then I moved it to run-pass :man_shrugging: 

Also, should I add tests for all methods I made `const fn`s?

Comment (oli-obk) [reply src/test/run-pass/const-endianess.rs]: More tests are always great ;)

run-pass is the right place for it

File: src/test/run-pass/const-endianess.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+const BE_U32: u32 = 55u32.to_be();
+const LE_U32: u32 = 55u32.to_le();
+const BE_U128: u128 = 999999u128.to_be();
+const LE_I128: i128 = -999999i128.to_le();
+
+fn main() {
+    assert_eq!(BE_U32, b(55u32).to_be());
+    assert_eq!(LE_U32, b(55u32).to_le());
+    assert_eq!(BE_U128, b(999999u128).to_be());
```
Comment (oli-obk) [inline src/test/run-pass/const-endianess.rs]: You should ignore this test on asm-js, because it doesn't support 128 bit integers

**Comment by bjorn3** [comment]
What happened here? Bors tried to test this while a rollup which contained this pr was already merged.

cc @alexcrichton

**Comment by Cyres** [comment]
@oli-obk In this PR, some tests are missing. With `ui` tests, the bug, which was fixed in #53699, would have been discovered way earlier.

**Pull request merged**